api/v1alpha1/local_object_reference_types.go

```diff
@@ -0,0 +1,110 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// ServicesDefaultLocalObjectReference identifies an API object within a
+// known namespace that defaults resource to services if unspecified.
+type ServicesDefaultLocalObjectReference struct {
+	// Group is the group of the referent.  Omitting the value or specifying
+	// the empty string indicates the core API group.  For example, use the
+	// following to specify a service:
+	//
+	// fooRef:
+	//   resource: services
+	//   name: myservice
+	//
+	// Otherwise, if the core API group is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +kubebuilder:default=core
+	// +optional
+	Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
+	// Resource is the API resource name of the referent. Omitting the value
+	// or specifying the empty string indicates the services resource. For example,
+	// use the following to specify a services resource:
+	//
+	// fooRef:
+	//   name: myservice
+	//
+	// Otherwise, if the services resource is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +kubebuilder:default=services
+	// +optional
+	Resource string `json:"resource" protobuf:"bytes,2,opt,name=resource"`
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	// +required
```

Comment:
Does `+required` do anything? In k/k I believe fields are required by default unless there is an `+optional` tag.

Reply:
I added it to be explicit and I see `+required` defined in other areas of the APIs. I'll remove `+required` and also submit a different PR to remove it elsewhere for consistency.